File: skellig-test-step-processing/src/main/java/org/skellig/teststep/processing/converter/DefaultTestDataConverter.java
package org.skellig.teststep.processing.converter;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class DefaultTestDataConverter implements TestDataConverter {

    private List<TestDataConverter> testDataConverters;

    protected DefaultTestDataConverter(List<TestDataConverter> valueConverters) {
        this.testDataConverters = valueConverters;
    }

    @Override
    public Object convert(Object value) {
        Object result = value;
        for (TestDataConverter testDataConverter : testDataConverters) {
            if (result == value) {
                result = testDataConverter.convert(result);
            } else {
                break;
            }
        }
        return result;
    }

    public static class Builder {

        private List<TestDataConverter> testDataConverters;
        private ClassLoader classLoader;

        public Builder() {
            testDataConverters = new ArrayList<>();
        }

        public Builder withClassLoader(ClassLoader classLoader) {
            this.classLoader = classLoader;
            return this;
        }

        public Builder withTestDataConverter(TestDataConverter testDataConverter) {
            this.testDataConverters.add(testDataConverter);
            return this;
        }

        public TestDataConverter build() {
            Objects.requireNonNull(classLoader, "ClassLoader must be provided");

            CsvTestDataConverter csvTestDataConverter = new CsvTestDataConverter(classLoader);
            testDataConverters.add(new TemplateTestDataConverter(classLoader, csvTestDataConverter));
            testDataConverters.add(csvTestDataConverter);

            return new DefaultTestDataConverter(testDataConverters);
        }
    }
}

File: skellig-test-step-processing/src/main/kotlin/org/skellig/teststep/processing/converter/DefaultTestDataConverter.kt
package org.skellig.teststep.processing.converter

import java.util.*

class DefaultTestDataConverter private constructor(private val testDataConverters: List<TestDataConverter>) : TestDataConverter {

    override fun convert(testData: Any?): Any? {
        var result = testData
        for (testDataConverter in testDataConverters) {
            if (result === testData) {
                result = testDataConverter.convert(result)
            } else {
                break
            }
        }
        return result
    }

    class Builder {

        private val testDataConverters: MutableList<TestDataConverter> = arrayListOf()
        private var classLoader: ClassLoader? = null

        fun withClassLoader(classLoader: ClassLoader?) = apply { this.classLoader = classLoader }

        fun withTestDataConverter(testDataConverter: TestDataConverter) = apply { testDataConverters.add(testDataConverter) }

        fun build(): TestDataConverter {
            Objects.requireNonNull(classLoader, "ClassLoader must be provided")

            val testDataFromCsvConverter = TestDataFromCsvConverter(classLoader!!)
            testDataConverters.add(TestDataToJsonConverter())
            testDataConverters.add(TestDataFromFTLConverter(classLoader!!, testDataFromCsvConverter))
            testDataConverters.add(TestDataFromIfStatementConverter())
            testDataConverters.add(testDataFromCsvConverter)
            testDataConverters.add(TestDataToBytesConverter())

            return DefaultTestDataConverter(testDataConverters)
        }
    }
}